fdb-extensions/src/test/java/com/apple/foundationdb/async/rabitq/FhtKacRotatorTest.java

```diff
@@ -0,0 +1,133 @@
+import com.apple.foundationdb.async.hnsw.RealVectorTest;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ObjectArrays;
+import com.google.common.collect.Sets;
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import javax.annotation.Nonnull;
+import java.util.Random;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+public class FhtKacRotatorTest {
+    @Nonnull
+    private static Stream<Arguments> randomSeedsWithDimensionality() {
+        return Sets.cartesianProduct(ImmutableSet.of(3, 5, 10, 128, 768, 1000))
+                .stream()
+                .flatMap(arguments ->
+                        LongStream.generate(() -> new Random().nextLong())
+                                .limit(3)
+                                .mapToObj(seed -> Arguments.of(ObjectArrays.concat(seed, arguments.toArray()))));
+    }
+
+    @ParameterizedTest(name = "seed={0} dimensionality={1}")
+    @MethodSource("randomSeedsWithDimensionality")
+    void testSimpleTest(final long seed, final int dimensionality) {
+        final FhtKacRotator rotator = new FhtKacRotator(seed, dimensionality, 10);
+
+        final Random random = new Random(seed);
+        final RealVector x = RealVectorTest.createRandomDoubleVector(random, dimensionality);
+
+        final RealVector y = rotator.operate(x);
+        final RealVector z = rotator.operateTranspose(y);
+
+        // Verify ||x|| ≈ ||y|| and P^T P ≈ I
+        double nx = norm2(x);
+        double ny = norm2(y);
+        double maxErr = maxAbsDiff(x, z);
+        System.out.printf("||x|| = %.6f  ||Px|| = %.6f  max|x - P^T P x|=%.3e%n", nx, ny, maxErr);
+    }
+
+    @Test
+    void testRotationIsStable() {
+        final FhtKacRotator rotator1 = new FhtKacRotator(0, 128, 10);
+        final FhtKacRotator rotator2 = new FhtKacRotator(0, 128, 10);
+        Assertions.assertThat(rotator1).isEqualTo(rotator2);
+
+        final Random random = new Random(0);
+        final RealVector x = RealVectorTest.createRandomDoubleVector(random, 128);
+        final RealVector x_ = rotator1.operate(x);
+        final RealVector x__ = rotator2.operate(x);
+
+        Assertions.assertThat(x_).isEqualTo(x__);
+    }
+
+    @ParameterizedTest(name = "seed={0} dimensionality={1}")
+    @MethodSource("randomSeedsWithDimensionality")
+    void testOrthogonality(final long seed, final int dimensionality) {
+        final FhtKacRotator rotator = new FhtKacRotator(seed, dimensionality, 10);
+        final ColumnMajorRealMatrix p = new ColumnMajorRealMatrix(rotator.computeP().transpose().getData());
+
+        for (int j = 0; j < dimensionality; j ++) {
+            final RealVector rotated = rotator.operateTranspose(new DoubleRealVector(p.getColumn(j)));
```

Comment:
Is it worth testing that rotating the rows of `rotator.computeP()` also produces the identity matrix?